server/lib/sync-agent.js

```diff
@@ -127,15 +128,48 @@ export default class SyncAgent {
       last_updated_at: options.last_updated_at || oneDayAgo.toISOString(),
       import_start_date: moment().subtract(this.ship.private_settings.import_days, "days").format()
     };
+
     const wrappedQuery = this.adapter.in.wrapQuery(query, replacements);
-    // Run the method for the specific adapter.
+      // Run the method for the specific adapter.
     return this.adapter.in.runQuery(this.client, wrappedQuery, options)
       .then(result => {
         this.adapter.in.closeConnection(this.client);
+
+        const validationResult = this.validateRows(result.rows);
+        if (!validationResult.isValid) {
+          return { entries: result.rows, errors: validationResult.errors };
+        }
+
+        if (this.db_type === "postgres" && this.postgresResultIsJson(result.fields)) {
```

Comment:
I would delegate the responsibility of validation to the adapters if we have special cases from different dbs